File: tests/test-sorting-network.cpp
/*********************                                                        */
/*! \file test-disjointset.cpp
** \verbatim
** Top contributors (to current version):
**   Makai Mann
** This file is part of the smt-switch project.
** Copyright (c) 2020 by the authors listed in the file AUTHORS
** in the top-level source directory) and their institutional affiliations.
** All rights reserved.  See the file LICENSE in the top-level source
** directory for licensing information.\endverbatim
**
** \brief Tests for SortingNetwork.
**
**
**/

#include <gtest/gtest.h>

#include <utility>
#include <vector>

#include "available_solvers.h"
#include "smt.h"
#include "sorting_network.h"
#include "utils.h"

using namespace smt;
using namespace std;

namespace smt_tests {

class SortingNetworkTests
    : public ::testing::Test,
      public ::testing::WithParamInterface<SolverConfiguration>
{
 protected:
  void SetUp() override
  {
    solver = create_solver(GetParam());
    solver->set_opt("produce-models", "true");
    boolsort = solver->make_sort(BOOL);
    for (size_t i = 0; i < NUM_VARS; ++i)
    {
      boolvec.push_back(solver->make_symbol("b" + std::to_string(i), boolsort));
    }
  }
  SmtSolver solver;
  Sort boolsort;
  TermVec boolvec;
  size_t NUM_VARS = 8;
};

TEST_P(SortingNetworkTests, TestSortingNetwork)
{
  SortingNetwork sn(solver);
  TermVec sorted = sn.sorting_network(boolvec);

  // Test each possible return value
  for (size_t num_true = 0; num_true <= NUM_VARS; ++num_true)
  {
    solver->push();
    if (num_true)
    {
      // ensure there are at least num_true set to true
      solver->assert_formula(sorted[num_true - 1]);
    }

    if (num_true < NUM_VARS)
    {
      // ensure there aren't more than num_true set to true
      solver->assert_formula(solver->make_term(Not, sorted[num_true]));
    }

    solver->check_sat();

    Term true_ = solver->make_term(true);
    size_t counted_true = 0;
    string model = "\n";
    Term val;
    for (const auto & bb : boolvec)
    {
      val = solver->get_value(bb);
      if (val == true_)
      {
        counted_true++;
      }
      model += "\t" + bb->to_string() + " := " + val->to_string() + "\n";
    }

    EXPECT_EQ(num_true, counted_true)
        << "Incorrect sorting network for model " + model;

    solver->pop();
  }
}

INSTANTIATE_TEST_SUITE_P(
    ParameterizedSolverSortingNetworkTests,
    SortingNetworkTests,
    testing::ValuesIn(available_non_generic_solver_configurations()));

}  // namespace smt_tests

File: tests/python/test_sorting_network.py
# SPDX-FileCopyrightText: 2021 the smt-switch authors
# SPDX-FileContributor: Makai Mann
# SPDX-License-Identifier: BSD-3-Clause

"""Test SortingNetwork through Python bindings.

See include/sorting_network.h for more information on the SortingNetwork class.
"""

from itertools import product

import pytest

import smt_switch as ss


@pytest.mark.parametrize(
    ("create_solver", "num_vars"), list(product(ss.solvers.values(), [3, 6, 8]))
)
def test_sorting_network(create_solver, num_vars):
    solver = create_solver(logging=False)
    solver.set_opt("produce-models", "true")
    solver.set_opt("incremental", "true")

    boolsort = solver.make_sort(ss.sortkinds.BOOL)
    boollist = [solver.make_symbol("b" + str(i), boolsort) for i in range(num_vars)]

    sn = ss.SortingNetwork(solver)
    sortedlist = sn.sorting_network(boollist)

    # Test each possible return value
    for num_true in range(num_vars + 1):
        solver.push()
        if num_true:
            # ensure there are at least num_true set to true
            solver.assert_formula(sortedlist[num_true - 1])
        if num_true < num_vars:
            # ensure there aren't more than num_true set to true
            solver.assert_formula(
                solver.make_term(ss.primops.Not, sortedlist[num_true])
            )
        res = solver.check_sat()
        assert res.is_sat()

        # The boolean here is the term's value, not a flag, so there is no
        # keyword form that would read better.
        true_ = solver.make_term(True)  # noqa: FBT003
        counted_true = 0
        for bb in boollist:
            val = solver.get_value(bb)
            if val == true_:
                counted_true += 1
        assert counted_true == num_true
        solver.pop()
